### PR TITLE
[REG2.066] Issue 14886 - std.parallelism.parallel with large static array seems to hang compile

### PR DESCRIPTION
**Pull request by 9rnsr**
https://issues.dlang.org/show_bug.cgi?id=14886

### CODE REVIEW

**Comment by yebblies** [comment]
This PR is over 60% code style changes.  This is the kind of formatting churn that I'm concerned about.  I understand the need for better source formatting, but we need the format agreed on and automated, otherwise it will never end.  These changes really don't belong in a regression fix PR.

**Comment by 9rnsr** [comment]
It's code style regression which had been caused by magicport. I'm just fixing it.

**Comment by yebblies** [comment]
Arbitrarily reformatting code is not 'fixing it'.  It's just creating churn.

**Comment by 9rnsr** [comment]
* In other PR, I was recommended by Walter to convert C style `for` to D style `foreach`.
* Line spaces and wrap long expression are revived from CDMD code.

**Comment by yebblies** [comment]
And what do either of those have to do with Issue 14886?

**Comment by WalterBright** [comment]
I agree that formatting changes belongs in separate PRs. Please do not combine them. Short PRs are much easier to review. The same goes for refactoring changes. Separate them, please.

**Comment by 9rnsr** [comment]
I like to clean code at first, because it will clarify the problematic point.

The cleanup commit and bugfixes are actually separated. I think it won't block reviewing.

**Comment by WalterBright** [comment]
This PR has 465 lines of change. Complex PRs are hard enough to review without lumping in extensive refactorings. Please make them separate PRs. We've discussed this before.

It's not just reviewing the individual commits. It's:

1. The "take it or leave it" of all or nothing.

2. If a bug is introduced, the git bisect leads to giant diffs as the culprit, again, making things harder.

3. Explorer behaves miserably when github presents large diffs.

4. I have only so many lines on my monitor. Big changes are a big pain in the ass to review.

And sometimes, there are simply too many "clean up" refactorings going on, making it impossible to compare source code from one release to the next. This makes it really really hard to continue understanding how the code works. Stability has a virtue all its own, when lots of people have to work with the code.

**Comment by 9rnsr** [comment]
I removed the cleanup commit.

**Comment by MartinNowak** [comment]
We really should use dfmt at some point to avoid any formatting work an discussions.
I'd even be willing to live with the few issues dfmt still has.

**Comment by yebblies** [comment]
> We really should use dfmt at some point to avoid any formatting work an discussions.

Agreed.

> I'd even be willing to live with the few issues dfmt still has.

Mmm I'm not quite sure it's ready.

**Comment by 9rnsr** [comment]
@MartinNowak magicport had removed many line spaces for logical code block separators. It won't be revived by dfmt.

**Comment by 9rnsr** [comment]
@yebblies I removed all format changes in this PR. So, isn't this ready for review to you?

**Comment by yebblies** [comment]
Ok, so this avoids allocating thousands of elements in array literal default init, makes sense.

I'm concerned that this requires so many changes in so many places.  I think it can be improved by providing foreach iteration over the expanded set of elements, instead of doing this manually in several places.  Then most of the iterations would become `foreach (e; &ale.expandElements)`.

**Comment by MartinNowak** [comment]
Your approach makes sense, but it's very invasive and brittle.
We should abstract the 2 iteration schemes, e.g. into a foreach construct or a simple "iterator" w/ a next() method. Also the name basis is somewhat unclear, how about defaultValue or so.

File: src/dcast.d [hunk 1]
```diff
@@ -677,12 +677,19 @@ extern (C++) MATCH implicitConvTo(Expression e, Type t)
...
+                        if (m < result)
+                            result = m;
+                    }
                     for (size_t i = 0; i < e.elements.dim; i++)
                     {
                         Expression el = (*e.elements)[i];
                         if (result == MATCHnomatch)
                             break;
-                        // no need to check for worse
+                        if (!el)
```
Comment (WalterBright) [inline src/dcast.d]: Why wasn't this check needed before?

File: src/expression.d [hunk 2]
```diff
@@ -4561,6 +4561,7 @@ public:
 extern (C++) final class ArrayLiteralExp : Expression
 {
 public:
+    Expression basis;
```
Comment (WalterBright) [inline src/expression.d]: What is `basis`? Please document.

File: src/expression.d [hunk 3]
```diff
@@ -4579,9 +4580,33 @@ public:
         elements.push(e);
     }
 
+    extern (D) this(Loc loc, Expression basis, Expressions* elements)
```
Comment (WalterBright) [inline src/expression.d]: When adding functions, please add Ddoc comment.

File: src/constfold.d [hunk 4]
```diff
@@ -770,6 +770,10 @@ extern (C++) UnionExp Equal(TOK op, Type type, Expression e1, Expression e2)
             {
                 Expression ee1 = (*es1.elements)[i];
                 Expression ee2 = (*es2.elements)[i];
+                if (!ee1)
+                    ee1 = es1.basis;
```
Comment (WalterBright) [inline src/constfold.d]: I see this pattern repeated dozens of times in this change. I suggest making a `get` function that retrieves element `i` instead.

**Comment by WalterBright** [comment]
> I removed the cleanup commit.

Thanks, Kenji!

File: src/dcast.d (hunk 1 above)
Comment (9rnsr) [inline src/dcast.d]: Because all elements in ArrayLiteralExp had been assumed to be non-null.

File: src/expression.d (hunk 2 above)
Comment (9rnsr) [inline src/expression.d]: I added explanation comment on that field.

File: src/expression.d (hunk 3 above)
Comment (9rnsr) [inline src/expression.d]: It will become obvious by the document on `basis` field.

File: src/constfold.d (hunk 4 above)
Comment (9rnsr) [inline src/constfold.d]: I added `getElement` function which returns `bases` when `(*elements)[i]` is `null`.

**Comment by 9rnsr** [comment]
Updated.

File: src/dcast.d (hunk 1 above)
Comment (WalterBright) [inline src/dcast.d]: were they non-null?

File: src/expression.d [hunk 5]
```diff
@@ -4609,6 +4628,47 @@ public:
...
 
+    final Expression getElement(size_t i)
+    {
+        auto el = (*elements)[i];
+        if (!el)
+            el = basis;
+        return el;
+    }
+
+    static Expressions* copyElements(Expression e1, Expression e2 = null)
```
Comment (WalterBright) [inline src/expression.d]: This could really use a Ddoc comment. What is `e2` for, for example?

File: src/dcast.d (hunk 1 above)
Comment (MartinNowak) [inline src/dcast.d]: Yes, but this PR leaves them null and lazily fills them w/ `basis` thereby avoid multiple semantic calls on the same element.

**Comment by MartinNowak** [comment]
> We should abstract the 2 iteration schemes, e.g. into a foreach construct or a simple "iterator" w/ a next() method.

Yeah, let's add 2 ranges for that, will make the code much clearer.
How about `sparseElements` and `denseElements`?
```d
extern (C++) static struct SparseElements
{
    size_t idx;
    Expression* defaultValue;
    Expressions values;

    bool empty()
    {
        return !defaultValue && idx == values.dim;
    }

    Expression* front()
    {
        return (!idx && defaultValue) ? defaultValue : values[idx];
    }

    void popFront()
    {
        if (defaultValue)
            defaultValue = null;
        else
            ++idx;
    }
}

extern (C++) static struct DenseElements
{
    size_t idx;
    Expression* defaultValue;
    Expressions values;

    bool empty()
    {
        return idx == values.dim;
    }

    Expression* front()
    {
        return values[idx] ? values[idx] : defaultValue;
    }

    void popFront()
    {
        ++idx;
    }
}
```

File: src/expression.d (hunk 5 above)
Comment (9rnsr) [inline src/expression.d]: Added ddoc comment.

**Comment by 9rnsr** [comment]
Rebased and resolved merge conflicts.

**Comment by WalterBright** [comment]
Thanks, Kenji!

**Comment by schuetzm** [comment]
Unfortunately, this introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=15123

**Pull request merged**